plugins/inlinetoolbar/plugin.js

```diff
@@ -168,13 +174,412 @@
+	 *
+	 * It's worth noting that priorities could be further customized by explicitly providing {@link CKEDITOR.plugins.inlinetoolbar.contextDefinition#priority},
+	 * so that it's possible to match a widget over a refresh callback.
+	 *
+	 * @class CKEDITOR.plugins.inlinetoolbar.contextManager
+	 * @constructor
+	 * @since 4.8
+	 * @param {CKEDITOR.editor} editor The editor instance which the toolbar is created for.
+	 */
+	function ContextManager( editor ) {
+		/**
+		 * Editor for which the manager was created for.
+		 *
+		 * @property {CKEDITOR.editor}
+		 */
+		this.editor = editor;
+
+		/**
+		 * List of contexts controlled by this manager.
+		 *
+		 * @private
+		 * @property {CKEDITOR.plugins.inlinetoolbar.context}
+		 */
+		this._contexts = [];
+
+		/**
+		 * Array of event listener references, created by the manager.
+		 *
+		 * @private
+		 * @property {Object[]} _listeners An array of objects returned by {@link CKEDITOR.event#on} method.
+		 */
+		this._listeners = [];
+
+		this._attachListeners();
+	}
+
+	ContextManager.prototype = {
+		/**
+		 * Adds a `context` to the tracked contexts list.
+		 *
+		 * @param {CKEDITOR.plugins.inlinetoolbar.context} context
+		 */
+		add: function( context ) {
+			this._contexts.push( context );
+		},
+
+		/**
+		 * Check each registered context against `selection` to find the best match. By default only one
+		 * toolbar per manager will be shown.
+		 *
+		 * @param {CKEDITOR.dom.selection/null} [selection=null] Selection to be used for probing toolbar. If none provided, a
+		 * _shrunk_ selection of current editor will be used.
+		 */
+		check: function( selection ) {
+			if ( !selection ) {
+				selection = this.editor.getSelection();
+
+				// Shrink the selection so that we're ensured innermost elements are available.
+				CKEDITOR.tools.array.forEach( selection.getRanges(), function( range ) {
+					range.shrink( CKEDITOR.SHRINK_ELEMENT, true );
+				} );
+			}
+
+			if ( !selection ) {
+				return;
+			}
+
+			var forEach = CKEDITOR.tools.array.forEach,
+				mainRange = selection.getRanges()[ 0 ],
+				path = mainRange && mainRange.startPath(),
+				highlightElement,
+				contextMatched;
+
+			// This function encapsulates matching algorithm.
+			function matchEachContext( contexts, matchingFunction, matchingArg1 ) {
+				forEach( contexts, function( curContext ) {
+					// Execute only if there's no picked context yet, or if probed context has a higher priority than
+					// currently matched one.
+					if ( !contextMatched || contextMatched.options.priority > curContext.options.priority ) {
+						var result = matchingFunction( curContext, matchingArg1 );
+
+						if ( !!result ) {
+							if ( result instanceof CKEDITOR.dom.element ) {
+								highlightElement = result;
+							} else {
+								// Reset what could have been previously set highlight element, as it's no longer relevant.
+								highlightElement = null;
+							}
+
+							contextMatched = curContext;
+						}
+					}
```

Comment:
Now as `_refresh*` functions API got simpler and returns `null` or element, couldn't this be simplified to:

```
if ( !contextMatched || contextMatched.options.priority > curContext.options.priority ) {
	highlightElement = matchingFunction( curContext, matchingArg1 );

	if ( !!highlightElement ) {
		contextMatched = curContext;
	}
}
```

?

Reply:
We can't simplify it that much as it would break test case when subsequent (higher priority) context returns `null` again. In that case it would cause `highlightElement` to be reset back to `null`.

Covered this with a test case in 8e89c0330836e0519bf6ff2f26e5185b13b5beb0.

Nonetheless we could simplify it now a bit, which I did in e382734c8.